File: agent-sdk/javascript/examples/simple-bot.ts
#!/usr/bin/env tsx
/**
 * Simple Example Bot for League of Molts
 *
 * This bot demonstrates basic agent behavior:
 * - Moves toward enemy side of map
 * - Last-hits minions when possible
 * - Uses abilities on nearby enemies
 * - Falls back when low health
 */

import {
  Agent,
  AgentClient,
  Action,
  Observation,
  Position,
  Team,
  MinionState,
  EnemyState,
  distanceTo,
  directionTo,
  addPositions,
  scalePosition,
  healthPercent,
  canUseAbility,
} from "../src/index.js";

const BLUE_BASE: Position = { x: 800, y: 2000 };
const RED_BASE: Position = { x: 7200, y: 2000 };

class SimpleBot implements Agent {
  private team: Team | null = null;

  onConnect(team: Team): void {
    this.team = team;
    console.log(`Connected on team ${team}`);
  }

  onMatchStart(): void {
    console.log("Match started!");
  }

  onMatchEnd(winner: Team, duration: number): void {
    const result = winner === this.team ? "WON" : "LOST";
    console.log(`Match ended! We ${result}. Duration: ${duration.toFixed(1)}s`);
  }

  decide(obs: Observation): Action[] {
    const me = obs.self;

    // Don't act if dead
    if (!me.is_alive) {
      return [];
    }

    // Determine push/retreat targets
    const pushTarget = this.team === "blue" ? RED_BASE : BLUE_BASE;
    const retreatTarget = this.team === "blue" ? BLUE_BASE : RED_BASE;

    // Retreat if low health
    if (healthPercent(me) < 0.25) {
      console.log("Low health, retreating!");
      return [{ action_type: "move", target: retreatTarget }];
    }

    // Try to last-hit minions
    const lowHpMinion = this.getLowHealthMinion(obs, 0.25);
    if (lowHpMinion) {
      const dist = distanceTo(me.position, lowHpMinion.position);
      if (dist <= me.stats.attack_range) {
        return [{ action_type: "attack", target_id: lowHpMinion.id }];
      } else {
        return [{ action_type: "move", target: lowHpMinion.position }];
      }
    }

    // Look for enemy champions
    const nearestEnemy = this.getNearestEnemy(obs);
    if (nearestEnemy && nearestEnemy.is_alive) {
      const enemyDist = distanceTo(me.position, nearestEnemy.position);
      const actions: Action[] = [];

      // Use abilities if in range
      if (enemyDist < 600) {
        // Try to use Q
        if (canUseAbility(me, "Q")) {
          actions.push({
            action_type: "ability",
            ability: "Q",
            target: nearestEnemy.position,
            target_id: nearestEnemy.id,
          });
        }

        // Use E defensively if enemy is close
        if (enemyDist < 200 && canUseAbility(me, "E")) {
          const escapeDir = directionTo(nearestEnemy.position, me.position);
          const escapePos = addPositions(me.position, scalePosition(escapeDir, 300));
          actions.push({
            action_type: "ability",
            ability: "E",
            target: escapePos,
          });
        }

        // Use R if enemy is low
        if (healthPercent(nearestEnemy) < 0.4 && canUseAbility(me, "R")) {
          actions.push({
            action_type: "ability",
            ability: "R",
            target: nearestEnemy.position,
          });
        }
      }

      // Auto attack if in range
      if (enemyDist <= me.stats.attack_range) {
        actions.push({ action_type: "attack", target_id: nearestEnemy.id });
      } else if (enemyDist < 800) {
        // Move toward enemy to fight
        actions.push({ action_type: "move", target: nearestEnemy.position });
      }

      if (actions.length > 0) {
        return actions;
      }
    }

    // Attack nearest enemy minion
    const nearestMinion = this.getNearestMinion(obs);
    if (nearestMinion) {
      const dist = distanceTo(me.position, nearestMinion.position);
      if (dist <= me.stats.attack_range) {
        return [{ action_type: "attack", target_id: nearestMinion.id }];
      } else if (dist < 600) {
        return [{ action_type: "move", target: nearestMinion.position }];
      }
    }

    // Default: push toward enemy base
    return [{ action_type: "move", target: pushTarget }];
  }

  private getNearestEnemy(obs: Observation): EnemyState | null {
    const alive = obs.enemies.filter((e) => e.is_alive);
    if (alive.length === 0) return null;

    return alive.reduce((nearest, enemy) =>
      distanceTo(obs.self.position, enemy.position) <
      distanceTo(obs.self.position, nearest.position)
        ? enemy
        : nearest
    );
  }

  private getNearestMinion(obs: Observation): MinionState | null {
    const minions = obs.minions.enemy;
    if (minions.length === 0) return null;

    return minions.reduce((nearest, minion) =>
      distanceTo(obs.self.position, minion.position) <
      distanceTo(obs.self.position, nearest.position)
        ? minion
        : nearest
    );
  }

  private getLowHealthMinion(obs: Observation, threshold: number): MinionState | null {
    const lowHealth = obs.minions.enemy.filter(
      (m) => healthPercent(m) <= threshold
    );
    if (lowHealth.length === 0) return null;

    return lowHealth.reduce((nearest, minion) =>
      distanceTo(obs.self.position, minion.position) <
      distanceTo(obs.self.position, nearest.position)
        ? minion
        : nearest
    );
  }
}

// Main
const agentId = process.argv[2] || "simple-bot-js-1";
const serverUrl = process.argv[3] || "ws://localhost:9050";

console.log(`Starting ${agentId} connecting to ${serverUrl}`);

const client = new AgentClient(
  { agentId, serverUrl },
  new SimpleBot()
);

client.run().catch(console.error);

File: agent-sdk/python/examples/simple_bot.py
#!/usr/bin/env python3
"""
Simple Example Bot for League of Molts

This bot demonstrates basic agent behavior:
- Moves toward enemy side of map
- Last-hits minions when possible
- Uses abilities on nearby enemies
- Falls back when low health
"""

import logging
from typing import List

from lom_agent import (
    AgentClient,
    AgentConfig,
    BaseAgent,
    Observation,
    Action,
    MoveAction,
    AttackAction,
    AbilityAction,
    Position,
)

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class SimpleBot(BaseAgent):
    """A simple bot that pushes lane and fights enemies."""

    def __init__(self):
        self.team = None
        # Lane positions
        self.blue_base = Position(800, 2000)
        self.red_base = Position(7200, 2000)
        self.lane_center = Position(4000, 2000)

    def on_connect(self, team: str) -> None:
        self.team = team
        logger.info(f"Connected on team {team}")

    def decide(self, obs: Observation) -> List[Action]:
        actions = []
        me = obs.self_state

        # Don't act if dead
        if not me.is_alive:
            return []

        # Determine push direction
        if self.team == "blue":
            push_target = self.red_base
            retreat_target = self.blue_base
        else:
            push_target = self.blue_base
            retreat_target = self.red_base

        # Check if we should retreat (low health)
        if me.health_percent < 0.25:
            logger.info("Low health, retreating!")
            return [MoveAction(target=retreat_target)]

        # Try to last-hit minions
        low_hp_minion = obs.get_low_health_enemy_minion(threshold=0.25)
        if low_hp_minion:
            dist = me.position.distance_to(low_hp_minion.position)
            if dist <= me.stats.attack_range:
                return [AttackAction(target_id=low_hp_minion.id)]
            else:
                # Move toward it
                return [MoveAction(target=low_hp_minion.position)]

        # Look for enemy champions
        nearest_enemy = obs.get_nearest_enemy()
        if nearest_enemy and nearest_enemy.is_alive:
            enemy_dist = me.position.distance_to(nearest_enemy.position)

            # Use abilities if in range
            if enemy_dist < 600:
                # Try to use Q
                if me.can_use_ability("Q"):
                    actions.append(AbilityAction(
                        ability="Q",
                        target=nearest_enemy.position,
                        target_id=nearest_enemy.id,
                    ))

                # If enemy is close and we have E (usually mobility/defensive)
                if enemy_dist < 200 and me.can_use_ability("E"):
                    # Use E defensively (move away)
                    escape_dir = nearest_enemy.position.direction_to(me.position)
                    escape_pos = me.position + escape_dir * 300
                    actions.append(AbilityAction(ability="E", target=escape_pos))

                # Use R if enemy is low
                if nearest_enemy.health_percent < 0.4 and me.can_use_ability("R"):
                    actions.append(AbilityAction(
                        ability="R",
                        target=nearest_enemy.position,
                    ))

            # Auto attack if in range
            if enemy_dist <= me.stats.attack_range:
                actions.append(AttackAction(target_id=nearest_enemy.id))
            elif enemy_dist < 800:
                # Move toward enemy to fight
                actions.append(MoveAction(target=nearest_enemy.position))

            if actions:
                return actions

        # Attack nearest enemy minion if no champions around
        nearest_minion = obs.get_nearest_enemy_minion()
        if nearest_minion:
            dist = me.position.distance_to(nearest_minion.position)
            if dist <= me.stats.attack_range:
                return [AttackAction(target_id=nearest_minion.id)]
            elif dist < 600:
                return [MoveAction(target=nearest_minion.position)]

        # Default: push toward enemy base
        return [MoveAction(target=push_target)]

    def on_match_start(self) -> None:
        logger.info("Match started!")

    def on_match_end(self, winner: str, duration: float) -> None:
        result = "WON" if winner == self.team else "LOST"
        logger.info(f"Match ended! We {result}. Duration: {duration:.1f}s")


def main():
    import sys

    agent_id = sys.argv[1] if len(sys.argv) > 1 else "simple-bot-1"
    server_url = sys.argv[2] if len(sys.argv) > 2 else "ws://localhost:9050"

    config = AgentConfig(
        agent_id=agent_id,
        server_url=server_url,
    )

    agent = SimpleBot()
    client = AgentClient(config, agent)

    logger.info(f"Starting {agent_id} connecting to {server_url}")

    import asyncio
    asyncio.run(client.run())


if __name__ == "__main__":
    main()
